### PR TITLE
Add converters to and from ImgLib2 types

### LINKED ISSUE #80: Support postprocessing of ImgLib2 types
Currently, when a module has outputs of `ComplexType` (and therefore `RealType` and `IntegerType`), a warning is logged such as:

```
[WARNING] Ignoring unsupported output: value [net.imglib2.type.numeric.integer.UnsignedByteType]
```

It would be nice to support the most common types by using their `getRealDouble()` or more specific methods to populate a results table. For "simple" types, this is already done in [`org.scijava.table.process.ResultsPostprocessor`](https://github.com/scijava/scijava-table/blob/2bcd3d8bcbc3146312c5b657aeb2f78ac907d908/src/main/java/org/scijava/table/process/ResultsPostprocessor.java#L73-L79).

I'm not sure how this is best achieved, as we cannot add an ImgLib2 dependency to `scijava-table`. But implementing a new `PostprocessorPlugin` in `imagej-common` would not allow to fill the same results table instance as the existing `scijava-table` plugin, so we'd need to re-implement the processing of simple types in a plugin of higher priority, or make the existing table somehow accessible to extend the behavior.

@ctrueden any suggestions, hints?

### PR DESCRIPTION
**Pull request by imagejan**
This PR resolves #80 (except for the changes in result post-processing which are tracked in https://github.com/scijava/scijava-table/issues/5).

Remaining tasks:
* [ ] the `NumberToNumericTypeConverters` by Leon Yang were just copied over from `imagej-server`; should we consider migrating their git history as well, @ctrueden ?
* [ ] implement more converters in both directions, i.e. for (Unsigned)ShortType, BitType, etc.
* [ ] the license headers should be added/updated

### CODE REVIEW

**Comment by imagejan** [comment]
A question to the Java generics experts (likely @ctrueden):

Is it possible to implement a `Converter<Integer, IntegerType<T>>` that can convert to *sub-classes* of `NumericType`? Such that when you call e.g. `convert(someInteger, UnsignedByteType.class)`, it would generate a `new UnsignedByteType(someInteger)`, and when you call `convert(someInteger, ShortType.class)`, it would create `new ShortType(someInteger)`, etc. for all implemented sub-classes of `IntegerType`?

Or do we need to implement an explicit converter for every ImgLib2 `*Type` as present for four of them currently in `imagej-server` (`Double`, `Float`, `Integer`, `Long`)?

I figure that for the other direction it should be simple, as long as we `convert(IntegerType, Integer)` with the help of `IntegerType.getInteger()`. So we probably need:

* `IntegerTypeToIntegerConverter` using `getInteger()`
* `BooleanTypeToBooleanConverter` using `get()`
* `RealTypeToDoubleConverter` using `ComplexType#getRealDouble()`
* maybe `RealTypeToFloatConverter` using `ComplexType#getRealFloat()` ?

What do you think?

**Comment by imagejan** [comment]
While writing a test for an unrelated task, I discovered that `org.scijava.convert.DefaultConverter` will handle conversion for:

> Objects where the destination Class has a constructor which takes that Object

... which is true for most of (all?) the `Number` to `*Type` conversions.

As such, the converters (written by @lnyng) that I wanted to migrate from `imagej-server` to here (in https://github.com/imagej/imagej-common/pull/82/commits/37fd588863d185209430503fa96b54cba4107c1d) are actually completely unnecessary (!).

The other direction (`NumericType -> Number`) remains valid and necessary.

I'll try to update this pull request soon (keeping the tests for conversion in both directions, and adding just the necessary converters).

**Comment by ctrueden** [comment]
@imagejan The eventual goal was to split up the `DefaultConverter` into pieces. And maybe consider whether that auto-wrapping constructor magic is actually wise... but since that's how things currently work, and we shouldn't break backwards compatibility by changing it, your plan (of not implementing the unnecessary conversion directions) sounds good to me! :+1:

**Comment by gselzer** [comment]
@imagejan I cleaned up your work in an effort to make progress on https://github.com/imagej/pyimagej/issues/162. 

Do you feel that it is necessary to keep [these five tests](https://github.com/imagej/imagej-common/blob/c5fc9d163273b1eff67757d23377111746392a82/src/test/java/net/imagej/convert/NumberAndTypeConversionTest.java#L30)? I guess they are special cases...

@ctrueden I rebased this over the latest master, and we have all passing tests. Can we get this merged quickly?

**Comment by gselzer** [comment]
I actually think I got the formatting setup correctly in 6a7f8f4. Does this commit's formatting look right to you @ctrueden?

**Comment by imagejan** [comment]
Thanks @gselzer for updating this PR!

> Do you feel that it is necessary to keep these five tests? I guess they are special cases...

It's not exactly necessary, but they don't hurt, right? Just some (common) special cases to make sure they work.

**Comment by gselzer** [comment]
> It's not exactly necessary, but they don't hurt, right? Just some (common) special cases to make sure they work.

Might as well keep them then, they don't hurt to keep.

File: src/test/java/net/imagej/convert/NumberAndTypeConversionTest.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+import net.imglib2.type.numeric.NumericType;
+import net.imglib2.type.numeric.real.DoubleType;
+import net.imglib2.type.numeric.real.FloatType;
+import net.imglib2.type.operators.ValueEquals;
+
+@RunWith(Parameterized.class)
+public class NumberAndTypeConversionTest {
+
+	@Parameters(name = "Type conversion test {index}: {0} <=> {1}")
```
Comment (ctrueden) [inline src/test/java/net/imagej/convert/NumberAndTypeConversionTest.java:28]: Heh, I didn't even know JUnit 4 supported parameterized testing. 🤷

**Pull request merged**